src/reduced_order/assemble_ECSW_residual.cpp

```diff
@@ -0,0 +1,162 @@
+
+        // Loop through the elements
+        for (const auto &cell : this->dg->dof_handler.active_cell_iterators())
+        {
+            int cell_num = cell->active_cell_index();
+            const int fe_index_curr_cell = cell->active_fe_index();
+            const dealii::FESystem<dim,dim> &current_fe_ref = this->dg->fe_collection[fe_index_curr_cell];
+            const int n_dofs_curr_cell = current_fe_ref.n_dofs_per_cell();
+
+            current_dofs_indices.resize(n_dofs_curr_cell);
+            cell->get_dof_indices(current_dofs_indices);
+
+            // Create L_e matrix for current cell
+            Epetra_Map LeRowMap(n_dofs_curr_cell, 0, epetra_comm);
+            Epetra_Map LeColMap(N_FOM_dim, 0, epetra_comm);
+            Epetra_CrsMatrix L_e(Epetra_DataAccess::Copy, LeRowMap, N_FOM_dim);
+            double posOne = 1.0;
+
+            for(int i = 0; i < n_dofs_curr_cell; i++){
+                const int col = current_dofs_indices[i];
+                L_e.InsertGlobalValues(i, 1, &posOne , &col);
+            }
+            L_e.FillComplete(LeColMap, LeRowMap);
+
+            // Extract residual contributions of the current cell into global dimension
+            Epetra_Vector local_r(LeRowMap);
+            Epetra_Vector global_r_e(LeColMap);
+            L_e.Multiply(false, epetra_right_hand_side, local_r);
+            L_e.Multiply(true, local_r, global_r_e);
+
+            // Find reduced-order representation of contribution
+            Epetra_Map cseRowMap(n_reduced_dim_POD, 0, epetra_comm);
+            Epetra_Vector c_se(cseRowMap);
+
+            epetra_test_basis->Multiply(true, global_r_e, c_se);
+            double *c_se_array = new double[n_reduced_dim_POD];
+
+            c_se.ExtractCopy(c_se_array);
+            
+            // Sub into entries of C and d
+            for (int k = 0; k < n_reduced_dim_POD; ++k){
+                int place = row_num+k;
+                C.InsertGlobalValues(place, 1, &c_se_array[k], &cell_num);
+                d.SumIntoGlobalValues(1, &c_se_array[k], &place);
+            }
+            
+        }
+        row_num+=n_reduced_dim_POD;
+        snap_num+=1;
+
+        // Check if number of training snapshots has been reached
+        if (this->all_parameters->hyper_reduction_param.num_training_snaps != 0) {
+            std::cout << "LIMITED NUMBER OF SNAPSHOTS"<< std::endl;
+            if (snap_num > (this->all_parameters->hyper_reduction_param.num_training_snaps-1)){
+                break;
+            }
+        }
+    }
+
+    C.FillComplete(ColMap, RowMap);
+
+    // Sub temp C and d into class A and b
+    this->A->reinit(C);
+    this->b.reinit(d.GlobalLength());
+    for(int z = 0 ; z < d.GlobalLength() ; z++){
+        this->b(z) = d[z];
+    }
+}
+
+#if PHILIP_DIM==1
+        template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM>;
+#endif
+
+#if PHILIP_DIM!=1
+        template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM+2>;
```

Comment:
```suggestion
    template class AssembleECSWRes<PHILIP_DIM, PHILIP_DIM+2>;
```

Reply:
removed extra tab
removed in assemble_ECSW_jacobian.cpp and assemble_ECSW_training_base.cpp as well